### PR TITLE
Unit tests for interacting with ECR (Resolver and Fetcher)

### PR DESCRIPTION
**Pull request by samuelkarp**
Add a bunch of unit tests covering interactions with ECR.  There's a lot here, but each commit should stand on its own and can be reviewed in sequence.

This PR currently covers the Resolver and Fetcher, but not the Pusher.  It can be reviewed now and Pusher tests can be submitted separately.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by samuelkarp** [comment]
cc @nmeyerhans @sipsma

I'd appreciate your review.  Thanks!

File: ecr/base.go [hunk 1]
```diff
@@ -32,21 +32,33 @@ var (
 )
 
 type ecrBase struct {
-	client  ecriface.ECRAPI
+	client  ecrAPI
 	ecrSpec ECRSpec
 }
 
+// ecrAPI contains only the ECR APIs that are called by the resolver
+type ecrAPI interface {
```
Comment (petderek) [inline ecr/base.go]: nit: link to godoc for ecrapi to help future code explorers

File: ecr/base.go [hunk 2]
```diff
@@ -32,21 +32,33 @@ var (
...
 	imageIdentifier := b.ecrSpec.ImageID()
 	log.G(ctx).WithField("imageIdentifier", imageIdentifier).Debug("ecr.base.manifest")
 	batchGetImageInput := &ecr.BatchGetImageInput{
-		RegistryId:         aws.String(b.ecrSpec.Registry()),
-		RepositoryName:     aws.String(b.ecrSpec.Repository),
-		ImageIds:           []*ecr.ImageIdentifier{imageIdentifier},
+		RegistryId:     aws.String(b.ecrSpec.Registry()),
+		RepositoryName: aws.String(b.ecrSpec.Repository),
+		ImageIds:       []*ecr.ImageIdentifier{imageIdentifier},
+		// TODO: Determine if this should be hard-coded
```
Comment (petderek) [inline ecr/base.go]: nit: functionally, it is. Might as well put it in a var () block.

File: ecr/fake_ecr_client_test.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	BatchCheckLayerAvailabilityFn func(aws.Context, *ecr.BatchCheckLayerAvailabilityInput, ...request.Option) (*ecr.BatchCheckLayerAvailabilityOutput, error)
+	InitiateLayerUploadFn         func(*ecr.InitiateLayerUploadInput) (*ecr.InitiateLayerUploadOutput, error)
+	UploadLayerPartFn             func(*ecr.UploadLayerPartInput) (*ecr.UploadLayerPartOutput, error)
+	CompleteLayerUploadFn         func(*ecr.CompleteLayerUploadInput) (*ecr.CompleteLayerUploadOutput, error)
+	PutImageFn                    func(aws.Context, *ecr.PutImageInput, ...request.Option) (*ecr.PutImageOutput, error)
+}
+
+var _ ecrAPI = (*fakeECRClient)(nil)
+
+func (f *fakeECRClient) BatchGetImageWithContext(ctx aws.Context, arg *ecr.BatchGetImageInput, opts ...request.Option) (*ecr.BatchGetImageOutput, error) {
```
Comment (petderek) [inline ecr/fake_ecr_client_test.go]: One way to make the zero value useful here would be to automagically return empty / nil if the function isn't defined:

```
if f.Batch.GetImageFn == nil {
    return nil, nil // or &BatchGetImageOutput{}
} 
```

File: ecr/fetcher_test.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+)
+
+func TestFetchUnimplemented(t *testing.T) {
+	fetcher := &ecrFetcher{}
+	desc := ocispec.Descriptor{
+		MediaType: "never-implemented",
+	}
+	_, err := fetcher.Fetch(context.Background(), desc)
+	assert.Error(t, err)
+	assert.Equal(t, unimplemented, err)
```
Comment (petderek) [inline ecr/fetcher_test.go]: nit: You may be able to collapse this into [EqualError](https://godoc.org/github.com/stretchr/testify/assert#EqualError)

File: ecr/fetcher_test.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+func TestFetchManifest(t *testing.T) {
+	registry := "registry"
+	repository := "repository"
+	imageTag := "tag"
+	imageManifest := "image manifest"
+	fakeClient := &fakeECRClient{}
+	fetcher := &ecrFetcher{
+		ecrBase{
```
Comment (petderek) [inline ecr/fetcher_test.go]: nit: `go vet` would complain about this not having a map key.

File: ecr/fetcher_test.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+	// test all supported media types
+	for _, mediaType := range []string{
+		ocispec.MediaTypeImageManifest,
+		images.MediaTypeDockerSchema2Manifest,
+		images.MediaTypeDockerSchema1Manifest,
+	} {
+		t.Run(mediaType, func(t *testing.T) {
+			callCount := 0
+			fakeClient.BatchGetImageFn = func(_ aws.Context, input *ecr.BatchGetImageInput, _ ...request.Option) (*ecr.BatchGetImageOutput, error) {
```
Comment (petderek) [inline ecr/fetcher_test.go]: ok this is pretty cool

File: ecr/fake_ecr_client_test.go (hunk 3 above)
Comment (samuelkarp) [reply ecr/fake_ecr_client_test.go]: I'd prefer not.  I like the panic approach as it makes it very obvious that the fake hasn't been initialized.

File: ecr/fetcher_test.go (hunk 5 above)
Comment (samuelkarp) [reply ecr/fetcher_test.go]: It's an embedded object, not a map.

File: ecr/base.go (hunk 2 above)
Comment (samuelkarp) [reply ecr/base.go]: Yeah, this is a question for the future: it's hard-coded now (and I'm not changing that), but we need to evaluate whether that's the right thing to do or not.  I suspect that it isn't, but need to do some testing.

File: ecr/fetcher_test.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+				},
+				Repository: repository,
+				Object:     imageTag,
+			},
+		},
+	}
+
+	// test all supported media types
+	for _, mediaType := range []string{
+		ocispec.MediaTypeImageManifest,
```
Comment (sipsma) [inline ecr/fetcher_test.go]: nit: here and elsewhere, there's sets of media types that are reused as group in the actual code and test code. They could be vars like
```
var (
   manifestMediaTypes = []string{ocispec.MediaTypeImageManifest, images.MediaTypeDockerSchema2Manifest, etc.}
   foreignMediaTypes = []string{images.MediaTypeDockerSchema2LayerForeign, etc.}
)
```

Just to make updating them in the future easier if needed

File: ecr/base.go (hunk 1 above)
Comment (samuelkarp) [reply ecr/base.go]: Done

File: ecr/fetcher_test.go (hunk 4 above)
Comment (samuelkarp) [reply ecr/fetcher_test.go]: Done

File: ecr/fetcher_test.go [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+	}
+	resolver := &ecrResolver{
+		clients: map[string]ecrAPI{
+			"fake": fakeClient,
+		},
+	}
+	fetcher, err := resolver.Fetcher(context.Background(), ref)
+	require.NoError(t, err, "failed to create fetcher")
+	_, err = fetcher.Fetch(context.Background(), ocispec.Descriptor{MediaType: mediaType})
+	assert.Error(t, err)
```
Comment (sipsma) [inline ecr/fetcher_test.go]: nit: does it make sense to assert the err or the Cause was the expected one? Same for the other sad-path tests below

File: ecr/fetcher_test.go (hunk 7 above)
Comment (samuelkarp) [reply ecr/fetcher_test.go]: In the code these are used as case statements; unfortunately I don't think you can use a slice like that.

File: ecr/fetcher_test.go (hunk 8 above)
Comment (samuelkarp) [reply ecr/fetcher_test.go]: Done

**Comment by samuelkarp** [comment]
@sipsma 

> There were also a few other files you edited that seem to need an updated year in the license header.

Fixed

**Pull request merged**